### PR TITLE
Add build_rpath

### PR DESCRIPTION
**Pull request by jpakkane**
Build_rpath is the opposite of install_rpath. It is defined when in the build dir but not after install.

Also includes a simple check for people trying to set it manually. It remains permitted for now because of backwards compatibility.

### CODE REVIEW

**Comment by nirbheek** [comment]
How are people supposed to use this? Is this supposed to point to directories outside the source tree? If so, are users expected to edit the build files and point every library and executable call to their custom prefixes?

**Comment by jpakkane** [comment]
Currently it is just an enabler. People seem to want to do this so it's better that they do it via a supported interface than adding random linker flags that break silently. We could add something fancier (e.g. automatically add build_rpath to external dependencies that are in nonstandard locations).

File: docs/markdown/Release-notes-for-0.42.0.md [hunk 1]
```diff
@@ -58,3 +58,14 @@ Rust's [linkage reference][rust-linkage].
...
 simultaneously by passing `-Db_sanitize=address,undefined` to Meson.
+
+## Added build_rpath keyword argument
+
+You can specify `build_rpath : '/foo/bar'` in build targets and the
+given path will get added to the target's rpath in the build tree. It
+is removed during the install step.
+
+Meson will print a warning when the user tries to add an rpath linker
+flag manually. This is not recommended because having multiple rpath
```
Comment (nirbheek) [inline docs/markdown/Release-notes-for-0.42.0.md]: Maybe also mention what "manually" means. i.e., via link_args: or LDFLAGS, etc.

File: mesonbuild/build.py [hunk 2]
```diff
@@ -669,6 +670,10 @@ def process_kwargs(self, kwargs, environment):
         for i in self.link_args:
             if not isinstance(i, str):
                 raise InvalidArguments('Link_args arguments must be strings.')
+        for l in self.link_args:
+            if '-Wl,-rpath' in l:
```
Comment (nirbheek) [inline mesonbuild/build.py]: This is not reliable; people can use `-Xlinker` `-rpath` and other forms.

File: docs/markdown/Release-notes-for-0.42.0.md (hunk 1 above)
Comment (jpakkane) [inline docs/markdown/Release-notes-for-0.42.0.md]: Updated.

File: mesonbuild/build.py (hunk 2 above)
Comment (jpakkane) [inline mesonbuild/build.py]: Converted slightly stricter. We can always make it more strict in the future but this should get 99% of all cases.

**Pull request merged**